cci_cameroon/analysis/clustering_rumours/rumour_clustering_model_development.py

```diff
@@ -0,0 +1,1048 @@
+set2 = data[data.category_id == 1][["comment", "category_id"]][:10]
+set3 = data[data.category_id == 2][["comment", "category_id"]][:10]
+to_use2 = (
+    pd.concat([set1, set2, set3]).drop_duplicates().reset_index().drop("index", axis=1)
+)
+# to_use2 = to_use.reset_index().drop("index",axis=1)
+
+# %%
+# create word embeddings using the transformer model
+sentence_embeddings2 = model.encode(to_use2.comment)
+
+# %%
+indp2 = faiss.index_factory(
+    sentence_embeddings2.shape[1], "Flat", faiss.METRIC_INNER_PRODUCT
+)
+faiss.normalize_L2(sentence_embeddings2)  # to be used for inner product computation
+# setting the number of neighbors to consider for graph connectivity
+neighbors2 = 10
+indp2.train(sentence_embeddings2)
+indp2.add(sentence_embeddings2)
+similarity_distance, sim_positions = indp2.search(
+    sentence_embeddings2, neighbors2  # consider only the first n elements
+)
+
+# %%
+edges, weighted_edges = generate_edges(
+    to_use2, "comment", sim_positions, similarity_distance
+)
+
+# %%
+nodes = list(set(to_use2.comment))
+
+# %%
+NG = nx.Graph()
+NG.add_nodes_from(nodes)
+NG.add_edges_from(edges)
+NG.add_weighted_edges_from(weighted_edges)
+
+# %%
+nx.draw(NG)
+
+# %%
+sentence_embeddings400 = model.encode(data.comment)
+
+# %%
+
+# %%
+# we shuffle the data and split it up into equal data chunks. We usBeginning with a chunk
+shuffled = model_data.sample(frac=1)
+result = np.array_split(shuffled, 25)
+df_list = []  # to hold metrics for each group of data chosen
+actual_comments_list = []  # holds the lists of retained communities
+rejected_groups_list = (
+    []
+)  # holds the list of All rejected communities with comment count that meet the set threshold
+# we start experimenting with 204 data points.
+dt = pd.concat([result[0], result[1], result[2], result[3]]).reset_index()
+dt.drop(["level_0", "index"], axis=1, inplace=True)
+
+# %%
+retained_comment_lists, rejected_cluster_lists, df_st = get_metrics(
+    dt, "comment", model
+)
+
+# %%
+df_st
+
+# %%
+retained_comment_lists[0][1]
+
+# %%
+df_list.append(df_st)
+actual_comments_list.append(
+    retained_comment_lists
+)  # stores the groups that are retained
+rejected_groups_list.append(rejected_cluster_lists)  # holds rejected clusters.
+
+# %%
+for count in range(
+    4, len(result) - 15
+):  # to work with a maximum of 500 comments (just a reasonable size)
+    dt = pd.concat([dt, result[count]]).reset_index()
+    dt.drop(["level_0", "index"], axis=1, inplace=True)
+    c, r, d = get_metrics(dt, "comment", model)
+    df_list.append(d)
+    actual_comments_list.append(c)
+    rejected_groups_list.append(r)
+
+# %%
+# We inspect the metrics computed and the quality of the clusters produced. This helps to choose the number of neighbors to take forward
+df_all = pd.concat(df_list).reset_index().drop("index", axis=1)
+df_all.head(50)
+
+# %%
+# In this section, we manually inspect the clusters formed when different number of neighbors are considered.
```

Comment:
I wonder if there might be a better way to display this to the user? Its tricky as its manual checks but its hard to tell if you need to uncomment out the below or if they are no longer needed. Could the manual checks be done in a loop inside a function and the results saved in a dictionary which is then saved to a file? Or is there a way of visualising the results?